### DISCUSSION TITLE
Feedback: Higher-order resolver

### DISCUSSION THREAD

**Discussion by TinsFox** [mswjs/mswjs.io#307] [Q&A]
- [Open page](https://mswjs.io/docs/recipes/higher-order-resolver)

## Feedback

Hello msw Team,

I hope this message finds you well. I want to express my appreciation for the msw project, which I've been using in my own project. It's been a great tool, but I've encountered some challenges in my usage, primarily due to incomplete documentation.

I understand this might be a direct request, but I believe that providing comprehensive code examples can significantly enhance the user experience. In this regard, I'd like to highlight an issue I've encountered while trying to create an authentication resolver. The code example provided in the documentation is incomplete and leaves room for confusion as it uses placeholders without specifying what should be returned. This can be particularly bewildering for users who are less familiar with the technology, such as myself. Specifically, I'm unsure about the line of code `return resolver(input)` and where the input variable comes from.

<img width="1103" alt="image" src="https://github.com/mswjs/mswjs.io/assets/33956589/86aa1c70-8941-4f48-940f-7a39735aba04">

I have two requests to improve the msw documentation:

1. **Tutorial Chapter**: I suggest adding a tutorial chapter to the documentation that provides a step-by-step guide for users to learn how to use msw progressively. It would be especially beneficial if the tutorial is in TypeScript, given that msw is written in TypeScript.

2. **Completeness**: Please avoid omitting code in the examples. Omissions can lead to significant confusion, especially for users who are not yet familiar with msw.

I'd like to offer my assistance in co-writing the new tutorial documentation. However, I admit that I don't have a full understanding of msw at this point and may require some guidance. I'm eager to contribute to improving the msw documentation and enhancing the experience for all users.

Thank you for considering these suggestions. I look forward to your response and am willing to collaborate to make msw even more accessible and user-friendly.

**Comment by kettanaito** [comment]
Hey, @TinsFox. Thank you for your kind words and for your feedback. This is totally relevant and the documentation example is an oversight on my part. I will push the fix to the docs today. 

> Tutorial Chapter

We _do_ have a tutorial you can follow! It's our [Getting started](https://mswjs.io/docs/getting-started) tutorial. If you read through it, you will learn why it's structured the way it is. It's modular and step-by-step, with the root-level "Getting started" page is your roadmap directing you to further pages in the docs to read and then return to the "Getting started" page to continue. 

> It would be especially beneficial if the tutorial is in TypeScript, given that msw is written in TypeScript.

Indeed, this would be useful. Pull requests are welcome to implement the JS/TS switch in all/some code examples in the docs! 

> Completeness: Please avoid omitting code in the examples. Omissions can lead to significant confusion, especially for users who are not yet familiar with msw.

Once again, an oversight. I don't believe we omit any other parts in our code snippets. If we do, please let me know. Thanks. 

> I'd like to offer my assistance in co-writing the new tutorial documentation.

This is appreciated! At this stage, I'd kindly ask you to read the "Getting started" tutorial and share your feedback with me. Let's see how we can improve what's already there before writing something new.

**Reply by TinsFox** [reply]
Thanks for your reply, I can't wait to continue integrating msw!

**Comment by kettanaito** [comment]
The [Higher-order resolver](https://mswjs.io/docs/recipes/higher-order-resolver/) recipe has been updated to be complete!
